File: src/lib/graphql.ts
import type { ApiRequest } from "@/types";

export type GraphqlPayload = {
  query: string;
  variables?: Record<string, unknown>;
  operationName?: string;
};

export type GraphqlError = {
  message: string;
  path?: Array<string | number>;
  extensions?: Record<string, unknown>;
};

export type GraphqlResponse = {
  data?: unknown;
  errors?: GraphqlError[];
};

export function defaultGraphqlQuery(): string {
  return `query Example {\n  __typename\n}`;
}

export function defaultGraphqlVariables(): string {
  return "{}";
}

export function buildGraphqlBody(
  request: Pick<ApiRequest, "graphqlQuery" | "graphqlVariables" | "graphqlOperationName">,
): string {
  const query = request.graphqlQuery.trim();
  const variablesRaw = request.graphqlVariables.trim() || "{}";
  const variables = JSON.parse(variablesRaw) as unknown;

  if (variables !== null && typeof variables !== "object") {
    throw new Error("GraphQL variables must be a JSON object.");
  }

  const payload: GraphqlPayload = {
    query,
    variables: (variables ?? {}) as Record<string, unknown>,
  };

  const operationName = request.graphqlOperationName.trim();
  if (operationName) {
    payload.operationName = operationName;
  }

  return JSON.stringify(payload);
}

export function validateGraphqlRequest(
  request: Pick<ApiRequest, "graphqlQuery" | "graphqlVariables" | "graphqlOperationName">,
): string | null {
  if (!request.graphqlQuery.trim()) {
    return "GraphQL query is required.";
  }

  try {
    buildGraphqlBody(request);
    return null;
  } catch (error) {
    return error instanceof Error ? error.message : "Invalid GraphQL request.";
  }
}

export function parseGraphqlResponse(body: string): GraphqlResponse | null {
  try {
    const parsed = JSON.parse(body) as GraphqlResponse;
    if (!parsed || typeof parsed !== "object") return null;
    if (!("data" in parsed) && !("errors" in parsed)) return null;
    return parsed;
  } catch {
    return null;
  }
}

export function formatGraphqlResponse(body: string): string {
  const parsed = parseGraphqlResponse(body);
  if (!parsed) return body;

  const sections: string[] = [];

  if (parsed.errors?.length) {
    sections.push(
      "Errors:\n" +
        parsed.errors
          .map((error) => {
            const path = error.path?.length ? ` (path: ${error.path.join(".")})` : "";
            return `- ${error.message}${path}`;
          })
          .join("\n"),
    );
  }

  if (parsed.data !== undefined) {
    sections.push("Data:\n" + JSON.stringify(parsed.data, null, 2));
  }

  return sections.join("\n\n") || body;
}

export const GRAPHQL_INTROSPECTION_QUERY = `query PulseIntrospection {
  __schema {
    queryType { name }
    mutationType { name }
    types {
      kind
      name
      description
      fields {
        name
        description
        args { name }
        type { kind name ofType { kind name ofType { kind name ofType { kind name } } } }
      }
    }
  }
}`;

export type GraphqlTypeRef = {
  kind?: string;
  name?: string | null;
  ofType?: GraphqlTypeRef | null;
};

export type GraphqlField = {
  name: string;
  description?: string | null;
  args?: Array<{ name: string }>;
  type?: GraphqlTypeRef;
};

export type GraphqlType = {
  kind?: string;
  name?: string | null;
  description?: string | null;
  fields?: GraphqlField[] | null;
};

export type GraphqlSchema = {
  queryType?: { name?: string } | null;
  mutationType?: { name?: string } | null;
  types: GraphqlType[];
};

export function parseGraphqlSchema(body: string): GraphqlSchema | null {
  const parsed = parseGraphqlResponse(body);
  const schema = (parsed?.data as { __schema?: GraphqlSchema } | undefined)?.__schema;
  if (!schema?.types) return null;
  return schema;
}

export function formatGraphqlTypeRef(type: GraphqlTypeRef | undefined): string {
  if (!type) return "";
  if (type.kind === "NON_NULL") return `${formatGraphqlTypeRef(type.ofType ?? undefined)}!`;
  if (type.kind === "LIST") return `[${formatGraphqlTypeRef(type.ofType ?? undefined)}]`;
  return type.name ?? "";
}

export function visibleGraphqlTypes(schema: GraphqlSchema): GraphqlType[] {
  return schema.types
    .filter((type) => type.name && !type.name.startsWith("__") && (type.fields?.length ?? 0) > 0)
    .sort((left, right) => (left.name ?? "").localeCompare(right.name ?? ""));
}

File: python/pulse/graphql.py
"""GraphQL body + a compact introspection query (same shape as the desktop explorer)."""

from __future__ import annotations

import json
from typing import Any

INTROSPECTION_QUERY = """query PulseIntrospection {
  __schema {
    queryType { name }
    mutationType { name }
    types {
      kind
      name
      description
      fields {
        name
        description
        args { name }
        type { kind name ofType { kind name ofType { kind name ofType { kind name } } } }
      }
    }
  }
}"""


def build_body(query: str, variables: object = None, operation_name: str | None = None) -> str:
    payload: dict[str, Any] = {"query": query}
    if variables is None or variables == "":
        payload["variables"] = {}
    elif isinstance(variables, str):
        loaded = json.loads(variables) if variables.strip() else {}
        if loaded is not None and not isinstance(loaded, dict):
            raise ValueError("GraphQL variables must be a JSON object")
        payload["variables"] = loaded or {}
    elif isinstance(variables, dict):
        payload["variables"] = variables
    else:
        raise ValueError("GraphQL variables must be a JSON object")
    name = (operation_name or "").strip()
    if name:
        payload["operationName"] = name
    return json.dumps(payload)


def summarize_schema(body: str | dict) -> dict[str, Any] | None:
    parsed = body if isinstance(body, dict) else json.loads(body)
    schema = ((parsed or {}).get("data") or {}).get("__schema")
    if not isinstance(schema, dict):
        return None
    types = []
    for item in schema.get("types") or []:
        if not isinstance(item, dict):
            continue
        name = str(item.get("name") or "")
        fields = item.get("fields") or []
        if not name or name.startswith("__") or not fields:
            continue
        types.append(
            {
                "kind": item.get("kind"),
                "name": name,
                "fields": [field.get("name") for field in fields if isinstance(field, dict) and field.get("name")],
            }
        )
    return {
        "queryType": (schema.get("queryType") or {}).get("name"),
        "mutationType": (schema.get("mutationType") or {}).get("name"),
        "types": types,
    }
